File: NameSiloDynDns/Services/UpdateService.cs
using Microsoft.Extensions.Hosting;
using NameSiloDynDns.NameSilo;
using Serilog;
using System;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;

namespace NameSiloDynDns.Services
{
    public class UpdateService : BackgroundService
    {
        private Timer timer;
        private readonly ILogger logger;
        private readonly NameSiloRepository repository;
        private readonly HostToUpdate hostToUpdate;

        public UpdateService(NameSiloRepository repository, HostToUpdate hostToUpdate, ILogger logger)
        {
            this.logger = logger.ForContext<UpdateService>();
            this.repository = repository;
            this.hostToUpdate = hostToUpdate;
        }

        protected override Task ExecuteAsync(CancellationToken stoppingToken)
        {
            logger.Information("Update Service Starting");
            var checkingPeriod = hostToUpdate.UpdateTimeSpan;
            timer = new Timer(CheckForUpdate, stoppingToken, TimeSpan.Zero, checkingPeriod);

            return Task.CompletedTask;
        }

        public override Task StopAsync(CancellationToken cancellationToken)
        {
            timer.Change(Timeout.InfiniteTimeSpan, TimeSpan.Zero);
            logger.Information("Stopping Update Service");

            return base.StopAsync(cancellationToken);
        }

        private void CheckForUpdate(object state)
        {
            try
            {
                var stoppingToken = (CancellationToken)state;
                stoppingToken.ThrowIfCancellationRequested();
                logger.Debug("Update Service Checking for Change Public IP Address");

                var dnsRecordList = repository.GetDnsRecordList().GetAwaiter().GetResult();

                var hostDnsRecord = dnsRecordList.ResourceRecords
                    .FirstOrDefault(r => r.Host.Equals(hostToUpdate.Host, StringComparison.InvariantCultureIgnoreCase)
                        && r.Type == "A");

                if (hostDnsRecord == null)
                {
                    logger.Error("The host {Host} does not exist as an 'A' record", hostToUpdate.Host);
                    return;
                }

                var host = hostDnsRecord.Host;
                var hostIP = hostDnsRecord.Value;
                var publicIP = dnsRecordList.CallingIpAddress;

                if (!dnsRecordList.CallingIpAddress.Equals(IPAddress.Parse(hostIP)))
                {
                    var response = repository.UpdateHostIpAddress(new UpdateHostMessage
                    {
                        RecordID = hostDnsRecord.RecordID,
                        Host = host,
                        IPAddress = publicIP
                    }).GetAwaiter().GetResult();

                    logger.ForContext(propertyName: "UpdatedIP", true)
                        .Warning("New Public IP Address {PublicIP}. Successfully Updated the host {Host} from {HostIP}",
                        publicIP, host, hostIP);
                }
                else
                {
                    logger.ForContext(propertyName: "UpdatedIP", true)
                        .Information("Public IP {PublicIP} and Host IP {HostIP} are the same on the host {Host}",
                        publicIP, hostIP, host);
                }
            }
            catch (HttpRequestException ex)
            {
                logger.Error(ex, $"HTTP Error to NameSilo");
            }
            catch (NameSiloApiException ex)
            {
                logger.Error(ex, $"Namesilo API Error: Operation: {ex.Operation}; Code: {ex.Code}");
            }
            catch (Exception ex)
            {
                logger.Error(ex, "Generic Error Checking for Update");
            }
        }

        public override void Dispose()
        {
            timer.Dispose();
        }
    }
}

File: NameSiloDynDns/UpdateService.fs
module NameSiloDynDns.Services

open Microsoft.Extensions.Hosting
open NameSiloDynDns.NameSilo
open Serilog
open System
open System.Net
open System.Net.Http
open System.Threading
open System.Threading.Tasks

type UpdateService(repository: NameSiloRepository, hostToUpdate: HostToUpdate, logger: ILogger) =
    inherit BackgroundService()

    let logger = logger.ForContext<UpdateService>()
    let mutable timer = Unchecked.defaultof<Timer>

    override this.ExecuteAsync(stoppingToken) =
        logger.Information("Update Service Starting")
        let checkingPeriod = hostToUpdate.UpdateTimeSpan
        timer <- new Timer(this.CheckForUpdate, stoppingToken, TimeSpan.Zero, checkingPeriod)

        Task.CompletedTask

    override this.StopAsync(stoppingToken) =
        timer.Change(Timeout.InfiniteTimeSpan, TimeSpan.Zero)
        |> ignore

        logger.Information("Stopping Update Service")

        base.StopAsync(stoppingToken)

    member this.CheckForUpdate(state: obj) =
        try
            let stoppingToken = state :?> CancellationToken
            stoppingToken.ThrowIfCancellationRequested()
            logger.Debug("Update Service Checking for Change Public IP Address")

            let dnsRecordList =
                repository
                    .GetDnsRecordList()
                    .GetAwaiter()
                    .GetResult()

            dnsRecordList.ResourceRecords
            |> Array.tryFind (fun r ->
                r.Host.Equals(hostToUpdate.Host, StringComparison.InvariantCultureIgnoreCase)
                && r.Type = "A")
            |> function
                | None -> logger.Error("The host {Host} does not exist as an 'A' record", hostToUpdate.Host)
                | Some hostDnsRecord ->
                    let host = hostDnsRecord.Host
                    let hostIP = hostDnsRecord.Value
                    let publicIP = dnsRecordList.CallingIpAddress

                    if not (dnsRecordList.CallingIpAddress.Equals(IPAddress.Parse(hostIP))) then
                        let response =
                            repository
                                .UpdateHostIpAddress(
                                    { RecordID = hostDnsRecord.RecordID
                                      Host = host
                                      IPAddress = publicIP }
                                )
                                .GetAwaiter()
                                .GetResult()

                        logger
                            .ForContext("UpdatedIP", true)
                            .Warning(
                                "New Public IP Address {PublicIP}. Successfully Updated the host {Host} from {HostIP}",
                                publicIP,
                                host,
                                hostIP
                            )
                    else
                        logger
                            .ForContext("UpdatedIP", true)
                            .Information(
                                "Public IP {PublicIP} and Host IP {HostIP} are the same on the host {Host}",
                                publicIP,
                                hostIP,
                                host
                            )
        with
        | :? HttpRequestException as ex -> logger.Error(ex, "HTTP Error to NameSilo")
        | :? NameSiloApiException as ex ->
            logger.Error(ex, $"Namesilo API Error: Operation: {ex.operation}; Code: {ex.code}")
        | ex -> logger.Error(ex, "Generic Error Checking for Update")

    override _.Dispose() = timer.Dispose()
